extra_data/tests/utils/gendata.py

```diff
@@ -0,0 +1,127 @@
+import sys
+from argparse import ArgumentParser
+from pathlib import Path
+
+import h5py
+
+from ..utils import progress_bar
+
+__all__ = ["clone"]
+
+
+def progress(processed, total, *, show=True):
+    """Show progress information"""
+    if not show:
+        return
+
+    pbar = progress_bar(processed, total)
+    if sys.stderr.isatty():
+        # "\x1b[2K": delete whole line, "\x1b[1A": move up cursor
+        print("\x1b[2K\x1b[1A\x1b[2K\x1b[1A", file=sys.stderr)
+    print(pbar, file=sys.stderr)
+
+
+def _clone_file_structure(
+    h5file: Path, output: Path, *, run_data=False, control_data=False
+) -> None:
+    clone = h5py.File(output / h5file.name, "w")
+
+    def visitor(name, obj):
+        if isinstance(obj, h5py.Group):
```

Comment:
We're planning to start using HDF5 links soon to (temporarily) have two different names for the corrected data sources, while we transition to new names: https://git.xfel.eu/calibration/planning/-/issues/170

It would be useful if this tool could preserve that (the current code won't). Checking for soft links is fairly easy:

```python
link = original.get(name, getlink=True)
if isinstance(link, h5py.SoftLink):
    ...
```

Checking for hard links is trickier - normally everything is a hard link, so you're really checking for two links pointing to the same object. But I think we're more likely to go with soft links, so we can probably overlook that for now.

Reply:
ah, thanks for the info. I've added support for soft links.

Reply:
While testing copying the softlinks, I realised `visititems` do not visit softlinks (already reported to h5py [issues](https://github.com/h5py/h5py/issues/671)) so I'd have to make my own visitor to handle that case... or ignore soft links for now.

Reply:
Aha, good point. I think we can get around that by using the somewhat hidden [`group.id.links.visit()`](http://api.h5py.org/h5l.html#h5py.h5l.LinkProxy.visit) method, which visits the links below a group, rather than the objects which the links point to. But that's extra complexity, so let's leave that until we actually start making files that use it.